fdbclient/S3Client.actor.cpp

```diff
@@ -74,27 +77,212 @@ Reference<S3BlobStoreEndpoint> getEndpoint(const std::string& s3url,
+                                              std::string bucket,
+                                              std::string objectName,
+                                              std::string uploadID,
+                                              PartState* part,
+                                              std::string partData,
+                                              int maxRetries,
+                                              int retryDelayMs) {
+	state int attempts = 0;
+	// Create packet queue for the part data
+	state UnsentPacketQueue packets;
+
+	while (attempts < maxRetries) {
+		try {
+			// Verify MD5 before upload
+			std::string calculatedMD5 = HTTP::computeMD5Sum(partData);
+			if (part->md5 != calculatedMD5) {
+				TraceEvent(SevWarn, "S3ClientUploadPartMD5Mismatch")
+				    .detail("Bucket", bucket)
+				    .detail("Object", objectName)
+				    .detail("Part", part->partNumber)
+				    .detail("Expected", part->md5)
+				    .detail("Calculated", calculatedMD5);
+				throw io_error();
+			}
+			PacketWriter pw(packets.getWriteBuffer(partData.size()), nullptr, Unversioned());
+			pw.serializeBytes(partData);
+
+			state std::string etag = wait(endpoint->uploadPart(
+			    bucket, objectName, uploadID, part->partNumber, &packets, partData.size(), calculatedMD5));
+			part->etag = etag;
+			part->completed = true;
+			return Void();
+		} catch (Error& e) {
+			attempts++;
+			if (attempts >= maxRetries) {
+				throw;
+			}
+
+			TraceEvent(SevWarn, "S3ClientUploadPartRetry")
+			    .error(e)
+			    .detail("Bucket", bucket)
+			    .detail("Object", objectName)
+			    .detail("Part", part->partNumber)
+			    .detail("Attempt", attempts)
+			    .detail("MaxAttempts", maxRetries);
+
+			wait(delay(retryDelayMs / 1000.0));
+		}
+	}
+	throw operation_failed();
+}
+
+// Read a range of bytes from a file asynchronously.
+ACTOR static Future<std::string> readFileRange(std::string filepath, int64_t size, int64_t offset) {
+	state Reference<IAsyncFile> file;
+	try {
+		Reference<IAsyncFile> f = wait(
+		    IAsyncFileSystem::filesystem()->open(filepath, IAsyncFile::OPEN_READONLY | IAsyncFile::OPEN_UNCACHED, 0));
+		file = f;
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "S3ClientReadFileError")
+		    .error(e)
+		    .detail("Filepath", filepath)
+		    .detail("Offset", offset)
+		    .detail("Size", size);
+		throw io_error();
+	}
+
+	state std::string result(size, '\0');
+	try {
+		int bytesRead = wait(file->read(&result[0], size, offset));
+		if (bytesRead != size) {
+			throw io_error();
+		}
+		return result;
+	} catch (Error& e) {
+		TraceEvent(SevWarn, "S3ClientReadFileError")
+		    .error(e)
+		    .detail("Filepath", filepath)
+		    .detail("Offset", offset)
+		    .detail("Size", size);
+		throw io_error();
+	}
+}
+
 // Copy filepath to bucket at resource in s3.
 ACTOR static Future<Void> copyUpFile(Reference<S3BlobStoreEndpoint> endpoint,
```

Comment:
Do we want to add some timespan measurement here? for debugging performance

Reply:
Also, we may want to do some read checksum of the uploaded file after the upload for the consistency check. We can use a flag to set this functionality. We may want to add TODO(BulkLoad) here.